src/sage/structure/unique_representation.py

```diff
@@ -559,6 +559,23 @@ class that inherits from :class:`UniqueRepresentation`: By adding
 from sage.misc.fast_methods import WithEqualityById
 
 
+class WithEqualityByArgs(metaclass=ClasscallMetaclass):
+    @staticmethod
+    def __classcall__(cls, *args, **options):
+        instance = typecall(cls, *args, **options)
+        instance._reduction = (cls, args, options)
+        return instance
+
+    def __eq__(self, other):
```

Comment:
A lot of work has gone into the processing of construction arguments for `UniquePresentation` etc, because there may be things there that don't allow for hashing (working around that is always problematic). You'd be suffering similar problems here. Are you sure that the `_reduction` vector you define here is fit for purpose? The arguments passed in might not compare well and/or may not be hashable. You'll need some experience with this class to see how well it serves.

Reply:
I agree that I have too little experience here. I just copied what is done in UniqueRepresentation

Reply:
For all UR objects, the final inputs have needed to be hashable (this has caused me some headaches with things that are naturally `dict` inputs). So there isn't really much of a loss here for the hashing. We can also specify that for it to be hashable, the inputs need to all be hashable.